### PR TITLE
Add annotations and annotations parsing

### PR DESCRIPTION
**Pull request by Nuckerr**
Example:
```java
public class ExampleClass {
    @Getter
     @Config("ban-message")
    public String banMessage;
}
```
```java
public class ExamplePlugin extends JavaPlugin {
    public void onEnable() {
        Yaml config = new Yaml(...);
        config.annotateClass(new ExampleClass());
       this.getLogger().info(config.getBanMessage());
    }
}
```

### CODE REVIEW

**Comment by yoyosource** [comment]
Your example is illogical. You cannot get the ban message with `getBanMessage()` on the config Object as the method is not present. Did you mean:
```java
public class ExampleClass {
    @Getter
    @Config("ban-message")
    public String banMessage;
}
```
```
public class ExamplePlugin extends JavaPlugin {
    public void onEnable() {
        Yaml config = new Yaml(...);
        ExampleClass exampleClass = ExampleClass();
        config.annotateClass(exampleClass);
       this.getLogger().info(exampleClass.getBanMessage());
    }
}
```

**Comment by KotlinFactory** [comment]
Hey,

Thanks for contributing to SimplixStorage.

This pull request looks promising however you need to revert the "optimize-imports" commit as it does not align with our code style requirements.

Regards,
Leonhard

**Comment by Nuckerr** [comment]
> Your example is illogical. You cannot get the ban message with `getBanMessage()` on the config Object as the method is not present. Did you mean:
> 
> ```java
> public class ExampleClass {
>     @Getter
>     @Config("ban-message")
>     public String banMessage;
> }
> ```
> 
> ```java
> public class ExamplePlugin extends JavaPlugin {
>     public void onEnable() {
>         Yaml config = new Yaml(...);
>         ExampleClass exampleClass = ExampleClass();
>         config.annotateClass(exampleClass);
>         this.getLogger().info(exampleClass.getBanMessage());
>     }
> }
> ```

Yeah, my bad

**Comment by KotlinFactory** [comment]
Thank you so much for this PR. Will have a more detailed look soon but it looks good that we can merge this!

File: src/main/java/de/leonhard/storage/annotation/Config.java [hunk 1]
```diff
@@ -0,0 +1,12 @@
+package de.leonhard.storage.annotation;
+
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+@Retention(RetentionPolicy.RUNTIME)
+@Target({ElementType.FIELD})
+public @interface Config {
```
Comment (KotlinFactory) [inline src/main/java/de/leonhard/storage/annotation/Config.java:10]: The name config clashes with the other "Config" class in the project. This is kind of confusing. We need a better name for this. "ConfigPath" seems way more accurate to me.

File: src/test/java/de/leonhard/storage/ConfigTest.java [hunk 2]
```diff
@@ -75,10 +76,37 @@ void testSet() {
     Assertions.assertTrue(config.getBoolean("Test-Key-1"));
   }
 
+  @Test
```
Comment (KotlinFactory) [inline src/test/java/de/leonhard/storage/ConfigTest.java:79]: Great to see tests already being included. Amazing work!

**Review by KotlinFactory — CHANGES_REQUESTED** [review]
After those issues are addressed we can merge this.

File: src/main/java/de/leonhard/storage/annotation/Config.java (hunk 1 above)
Comment (Nuckerr) [reply src/main/java/de/leonhard/storage/annotation/Config.java]: Resolved in 81a473f

File: src/main/java/de/leonhard/storage/Config.java [hunk 3]
```diff
@@ -56,6 +58,34 @@ public Config(final File file) {
     super(file);
   }
 
+  public void annotateClass(Object classInstance) {
+    Class<?> clazz = classInstance.getClass();
+    try {
+      for (Field field : clazz.getFields()) {
+        if(field.isAnnotationPresent(ConfigPath.class)) {
```
Comment (yoyosource) [inline src/main/java/de/leonhard/storage/Config.java:65]: Isn't this really slow, first checking If the annotation is present and then getting it?

I myself would use:
```java
ConfigPath configPath = field.getAnnotation(ConfigPath.class);
if (configPath != null) {
    field.setAccessible(true);
    field.set(classInstance, this.get(configPath.value(), field.getType()));
}
```

File: src/main/java/de/leonhard/storage/Config.java [hunk 4]
```diff
@@ -56,6 +58,34 @@ public Config(final File file) {
...
+          field.setAccessible(true);
+          field.set(classInstance, this.get(field.getAnnotation(ConfigPath.class).value(), field.getType()));
+        }
+      }
+    }catch (IllegalAccessException e) {
+      e.printStackTrace();
```
Comment (yoyosource) [inline src/main/java/de/leonhard/storage/Config.java:71]: Just printing the exception is not that good in my opinion, maybe logging it properly or letting the caller decide how to react would be better practice (maybe not with an Exception but a RuntimeException?)

File: src/main/java/de/leonhard/storage/Config.java [hunk 5]
```diff
@@ -56,6 +58,34 @@ public Config(final File file) {
...
+    }
+  }
+
+  public void annotateClass(Object classInstance, String section) {
+    Class<?> clazz = classInstance.getClass();
+    try {
+      for (Field field : clazz.getFields()) {
+        if(field.isAnnotationPresent(ConfigPath.class)) {
```
Comment (yoyosource) [inline src/main/java/de/leonhard/storage/Config.java:79]: See above.

File: src/main/java/de/leonhard/storage/Config.java [hunk 6]
```diff
@@ -56,6 +58,34 @@ public Config(final File file) {
...
+          field.setAccessible(true);
+          field.set(classInstance, this.get(section + "." + field.getAnnotation(ConfigPath.class).value(), field.getType()));
+        }
+      }
+    }catch (IllegalAccessException e) {
+      e.printStackTrace();
```
Comment (yoyosource) [inline src/main/java/de/leonhard/storage/Config.java:85]: See above.

File: src/main/java/de/leonhard/storage/Config.java [hunk 7]
```diff
@@ -56,6 +58,34 @@ public Config(final File file) {
...
+          field.setAccessible(true);
+          field.set(classInstance, this.get(field.getAnnotation(ConfigPath.class).value(), field.getType()));
+        }
+      }
+    }catch (IllegalAccessException e) {
+      e.printStackTrace();
+    }
+  }
+
+  public void annotateClass(Object classInstance, String section) {
```
Comment (yoyosource) [inline src/main/java/de/leonhard/storage/Config.java:75]: I have an Idea to reduce code duplications:

You have:
```java
public void annotateClass(Object classInstance) {...}
public void annotateClass(Object classInstance, String section) {...}
```

And both just call:
```java
public void annotateClass(Object classInstance, UnaryOperator<String> elementSelector) {...}
```
You would pass the `configPath.value()` as parameter to the `elementSelector` and maybe you could even use a `BiFunction<String, Field, String>` so you could react on the field and so?

Comment (yoyosource) [reply src/main/java/de/leonhard/storage/Config.java:75]: This is just an Idea to future proof this Idea.

**Review by yoyosource — CHANGES_REQUESTED** [review]

File: src/main/java/de/leonhard/storage/Config.java (hunk 7 above)
Comment (Nuckerr) [reply src/main/java/de/leonhard/storage/Config.java:75]: I think this is what you meant
image

Comment (yoyosource) [reply src/main/java/de/leonhard/storage/Config.java:67]: Yes that is what I meant, but maybe using a BiFunction<String, Field, String> would be even more elegant.

File: src/main/java/de/leonhard/storage/Config.java [hunk 8]
```diff
@@ -56,6 +60,28 @@ public Config(final File file) {
     super(file);
   }
 
+  public void annotateClass(Object classInstance) {
+    this.annotateClass(classInstance, s -> "");
```
Comment (yoyosource) [inline src/main/java/de/leonhard/storage/Config.java:64]: The Idea here was to return s, if no prefixing should take place.
You can even use `Function.identitiy()` I think.

File: src/main/java/de/leonhard/storage/Config.java [hunk 9]
```diff
@@ -56,6 +60,28 @@ public Config(final File file) {
...
+  }
+
+  public void annotateClass(Object classInstance, String section) {
+    this.annotateClass(classInstance, s -> section + ".");
```
Comment (yoyosource) [inline src/main/java/de/leonhard/storage/Config.java:68]: For this you would use `s -> section + "." + s`

File: src/main/java/de/leonhard/storage/Config.java [hunk 10]
```diff
@@ -56,6 +60,28 @@ public Config(final File file) {
...
+  }
+
+  public void annotateClass(Object classInstance, UnaryOperator<String> elementSelector) {
+    Class<?> clazz = classInstance.getClass();
+    try {
+      for (Field field : clazz.getFields()) {
+        ConfigPath configPath = field.getAnnotation(ConfigPath.class);
+        if(configPath != null) {
+          field.set(classInstance, this.get(elementSelector.apply(configPath.value()) + configPath.value(), field.getType()));
```
Comment (yoyosource) [inline src/main/java/de/leonhard/storage/Config.java:77]: Than you can just use the single Operator and not use another concat here.

**Comment by KotlinFactory** [comment]
Merged, thanks.
We can implement the requested changes later.

**Pull request merged**